syntax = "proto3";

package sliverpb;

enum ProtocolType {
    HTTP = 0;
    Drive = 1;
    TCP = 2;
    UDP = 3;
    NamedPipe = 4;
}

enum ImplantType {
    Session = 0;
    Beacon = 1;
    Pivot = 2;
}

message DriveConfig {
    string ClientID = 1;
    string ClientSecret = 2;
    string RefreshToken = 3;
    string UserAgent = 4;
    string StartExtension = 5;
    string SendExtension = 6;
    string RecvExtension = 7;
    string RegisterExtension = 8;
    string CloseExtension = 9;
    uint32 PollInterval = 10;
}

message HttpConfig {
    repeated string PollPaths = 1;
    repeated string PollFiles = 2;
    repeated string SessionPaths = 3;
    repeated string SessionFiles = 4;
    repeated string ClosePaths = 5;
    repeated string CloseFiles = 6;
    string UserAgent = 7;
    string OtpSecret = 8;
    uint32 MinNumberOfSegments = 9;
    uint32 MaxNumberOfSegments = 10;
    uint32 PollInterval = 11;
    string URL = 12;
}

message PivotConfig {
    string BindAddress = 1;
}

message Config {
    string ServerPublicKey = 1;
    string PeerPublicKey = 2;
    string PeerPrivateKey = 3;
    string ServerMinisignPublicKey = 4;
    string SliverName = 5;
    string ConfigID = 6;
    string PeerAgePublicKeySignature = 7;
    uint64 EncoderNonce = 8;
    uint32 MaxConnectionErrors = 9;
    uint64 ReconnectInterval = 10;
    ProtocolType Protocol = 11;
    ImplantType Type = 12;
    repeated DriveConfig DriveConfigs = 13;
    repeated HttpConfig HttpConfigs = 14;
    repeated PivotConfig PivotConfigs = 15;
    bool Loot = 16;
    bool LootClipboard = 17;
    string Proxy = 18;
    string SliverPath = 19;
    string MainExecutable = 20;
}